src/router_ics_worker.erl

```diff
@@ -0,0 +1,290 @@
+    ?MODULE:start_link(Args#{port => erlang:list_to_integer(Port)});
+start_link(#{host := Host, port := Port} = Args) when is_list(Host) andalso is_integer(Port) ->
+    gen_server:start_link({local, ?SERVER}, ?SERVER, Args, []);
+start_link(_Args) ->
+    ignore.
+
+-spec add(list(binary())) -> ok.
+add(DeviceIDs) ->
+    gen_server:cast(?SERVER, {add, DeviceIDs}).
+
+-spec update(list(binary())) -> ok.
+update(DeviceIDs) ->
+    gen_server:cast(?SERVER, {update, DeviceIDs}).
+
+-spec remove(list(binary())) -> ok.
+remove(DeviceIDs) ->
+    gen_server:cast(?SERVER, {remove, DeviceIDs}).
+
+%% ------------------------------------------------------------------
+%% gen_server Function Definitions
+%% ------------------------------------------------------------------
+init(#{pubkey_bin := PubKeyBin, sig_fun := SigFun, host := Host, port := Port} = Args) ->
+    lager:info("~p init with ~p", [?SERVER, Args]),
+    {ok, _, SigFun, _} = blockchain_swarm:keys(),
+    Backoff = backoff:type(backoff:init(?BACKOFF_MIN, ?BACKOFF_MAX), normal),
+    self() ! ?CONNECT,
+    {ok, #state{
+        pubkey_bin = PubKeyBin,
+        sig_fun = SigFun,
+        host = Host,
+        port = Port,
+        conn_backoff = Backoff
+    }}.
+
+handle_call(_Msg, _From, State) ->
+    lager:warning("rcvd unknown call msg: ~p from: ~p", [_Msg, _From]),
+    {reply, ok, State}.
+
+handle_cast(_Msg, #state{conn = undefined} = State) ->
+    lager:warning("can't handle cast msg: ~p", [_Msg]),
+    {noreply, State};
+handle_cast(_Msg, #state{route_id = undefined} = State) ->
+    lager:warning("can't handle cast msg: ~p", [_Msg]),
+    {noreply, State};
+handle_cast(
+    {add, DeviceIDs},
+    #state{pubkey_bin = PubKeyBin, sig_fun = SigFun, conn = Conn, route_id = RouteID} = State
+) ->
+    Euis = fetch_device_euis(apis, DeviceIDs),
+    case euis_req(Conn, PubKeyBin, SigFun, RouteID, add_euis, Euis) of
+        true -> lager:info("added ~p", [Euis]);
+        false -> lager:warning("failed to add ~p", [Euis])
+    end,
+    {noreply, State};
+handle_cast(
+    {update, DeviceIDs},
+    #state{pubkey_bin = PubKeyBin, sig_fun = SigFun, conn = Conn, route_id = RouteID} = State
+) ->
+    CachedEuis = fetch_device_euis(cache, DeviceIDs),
+    APIEuis = fetch_device_euis(apis, DeviceIDs),
+    ToRemove = CachedEuis -- APIEuis,
+    case euis_req(Conn, PubKeyBin, SigFun, RouteID, remove_euis, ToRemove) of
+        true -> lager:info("removed ~p", [ToRemove]);
+        false -> lager:warning("failed to remove ~p", [ToRemove])
+    end,
+    ToAdd = APIEuis -- CachedEuis,
+    case euis_req(Conn, PubKeyBin, SigFun, RouteID, add_euis, ToAdd) of
+        true -> lager:info("added ~p", [ToAdd]);
+        false -> lager:warning("failed to add ~p", [ToAdd])
+    end,
+    {noreply, State};
+handle_cast(
+    {remove, DeviceIDs},
+    #state{pubkey_bin = PubKeyBin, sig_fun = SigFun, conn = Conn, route_id = RouteID} = State
+) ->
+    Euis = fetch_device_euis(cache, DeviceIDs),
```

Comment:
Should we also remove these devices from the cache?

Reply:
yep